core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java

```diff
@@ -72,30 +74,35 @@ private boolean refill() throws IOException {
   }
 
   @Override
-  public int read() throws IOException {
+  public synchronized int read() throws IOException {
     if (!refill()) {
       return -1;
     }
     return byteBuffer.get() & 0xFF;
   }
 
   @Override
-  public int read(byte[] b, int offset, int len) throws IOException {
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
     if (!refill()) {
       return -1;
     }
+    if ((offset | len | (offset + len) | (b.length - (offset + len))) < 0) {
```

Comment:
Oh I see. Well it's reasonable to keep it then, but I'd at least add a comment I think. I didn't recognize this at all until I thought about it for a good few minutes.